libwebrtc/src/native/desktop_capturer.rs

```diff
@@ -0,0 +1,214 @@
+    fn default() -> Self {
+        Self {
+            source_type: SourceType::Screen,
+            include_cursor: false,
+            #[cfg(target_os = "macos")]
+            allow_sck_system_picker: true,
+        }
+    }
+}
+
+impl DesktopCapturerOptions {
+    pub(crate) fn new(source_type: SourceType) -> Self {
+        Self { source_type, include_cursor: false, ..Default::default() }
+    }
+
+    pub(crate) fn with_cursor(mut self, include: bool) -> Self {
+        self.include_cursor = include;
+        self
+    }
+
+    #[cfg(target_os = "macos")]
+    pub(crate) fn with_sck_system_picker(mut self, allow_sck_system_picker: bool) -> Self {
+        self.allow_sck_system_picker = allow_sck_system_picker;
+        self
+    }
+
+    pub(crate) fn to_sys_handle(&self) -> sys_dc::ffi::DesktopCapturerOptions {
+        let source_type = match self.source_type {
+            SourceType::Screen => sys_dc::ffi::SourceType::Screen,
+            SourceType::Window => sys_dc::ffi::SourceType::Window,
+            SourceType::Generic => sys_dc::ffi::SourceType::Generic,
+        };
+        let mut sys_handle = sys_dc::ffi::DesktopCapturerOptions {
+            source_type,
+            include_cursor: self.include_cursor,
+            allow_sck_system_picker: false,
+        };
+        #[cfg(target_os = "macos")]
+        {
+            sys_handle.allow_sck_system_picker = self.allow_sck_system_picker;
+        }
+        sys_handle
+    }
+}
+
+pub struct DesktopCapturer {
+    pub(crate) sys_handle: UniquePtr<sys_dc::ffi::DesktopCapturer>,
+}
+
+impl DesktopCapturer {
+    pub fn new(options: DesktopCapturerOptions) -> Option<Self> {
+        let sys_handle = new_desktop_capturer(options.to_sys_handle());
+        if sys_handle.is_null() {
+            None
+        } else {
+            Some(Self { sys_handle })
+        }
+    }
+
+    pub fn capture_frame(&self) {
+        self.sys_handle.capture_frame();
+    }
+
+    pub fn start<T>(&mut self, callback: T)
+    where
+        T: FnMut(CaptureResult, DesktopFrame) + Send + 'static,
+    {
+        let pin_handle = self.sys_handle.pin_mut();
+        let callback = DesktopCallback::new(callback);
+        let callback_wrapper = sys_dc::DesktopCapturerCallbackWrapper::new(Box::new(callback));
+        pin_handle.start(Box::new(callback_wrapper));
+    }
+
+    pub fn select_source(&self, id: u64) -> bool {
+        self.sys_handle.select_source(id)
+    }
+
+    pub fn get_source_list(&self) -> Vec<CaptureSource> {
+        let mut sources = Vec::new();
+        let source_list = self.sys_handle.get_source_list();
+        for source in source_list.iter() {
+            sources.push(CaptureSource { sys_handle: source.clone() });
+        }
+        sources
+    }
+}
+
+pub struct DesktopFrame {
+    pub(crate) sys_handle: UniquePtr<sys_dc::ffi::DesktopFrame>,
+}
+
+impl DesktopFrame {
+    pub fn new(sys_handle: UniquePtr<sys_dc::ffi::DesktopFrame>) -> Self {
```

Comment:
Does this ever get constructed outside the _libwebrtc_ crate? If not, it should be marked `pub(crate)`.